dll/win32/kernel32/client/fiber.c

```diff
@@ -312,32 +312,88 @@ IsThreadAFiber(VOID)
 }
 
 /*
- * @unimplemented
+ * @implemented
  */
 DWORD
 WINAPI
 FlsAlloc(PFLS_CALLBACK_FUNCTION lpCallback)
 {
-   (void)lpCallback;
+    DWORD dwFlsIndex;
+    PPEB Peb = NtCurrentPeb();
+    PVOID *ppFlsSlots;
+
+    RtlAcquirePebLock();
 
-   UNIMPLEMENTED;
-   SetLastError(ERROR_CALL_NOT_IMPLEMENTED);
-   return FLS_OUT_OF_INDEXES;
+    ppFlsSlots = NtCurrentTeb()->FlsData;
+
+    if (!Peb->FlsCallback &&
+        !(Peb->FlsCallback = RtlAllocateHeap(RtlGetProcessHeap(), HEAP_ZERO_MEMORY,
+                                             FLS_MAXIMUM_AVAILABLE * sizeof(PVOID))))
+    {
+        SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+        dwFlsIndex = FLS_OUT_OF_INDEXES;
+    }
+    else
+    {
+        dwFlsIndex = RtlFindClearBitsAndSet(Peb->FlsBitmap, 1, 1);
+        if (dwFlsIndex != FLS_OUT_OF_INDEXES)
+        {
+            if (!ppFlsSlots &&
+                !(ppFlsSlots = RtlAllocateHeap(RtlGetProcessHeap(), HEAP_ZERO_MEMORY,
+                                               (FLS_MAXIMUM_AVAILABLE + 2) * sizeof(PVOID))))
+            {
+                RtlClearBits(Peb->FlsBitmap, dwFlsIndex, 1);
+                dwFlsIndex = FLS_OUT_OF_INDEXES;
+                SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+            }
+            else
+            {
+                if (!NtCurrentTeb()->FlsData)
+                    NtCurrentTeb()->FlsData = ppFlsSlots;
+
+                ppFlsSlots[dwFlsIndex + 2] = 0; /* clear the value */
+                Peb->FlsCallback[dwFlsIndex] = lpCallback;
+            }
+        }
+        else
+        {
+            SetLastError(ERROR_NO_MORE_ITEMS);
+        }
+    }
+    RtlReleasePebLock();
+    return dwFlsIndex;
 }
 
 
 /*
- * @unimplemented
+ * @implemented
  */
 BOOL
 WINAPI
 FlsFree(DWORD dwFlsIndex)
 {
-    (void)dwFlsIndex;
+    BOOL ret;
+    PPEB Peb = NtCurrentPeb();
+    PVOID *ppFlsSlots;
 
-    UNIMPLEMENTED;
-    SetLastError(ERROR_CALL_NOT_IMPLEMENTED);
-    return FALSE;
+    RtlAcquirePebLock();
+
+    ppFlsSlots = NtCurrentTeb()->FlsData;
+    ret = RtlAreBitsSet(Peb->FlsBitmap, dwFlsIndex, 1);
+    if (ret)
+    {
+        RtlClearBits(Peb->FlsBitmap, dwFlsIndex, 1);
+        /* FIXME: call Fls callback */
```

Comment:
So, something like: `if (Peb->FlsCallback[dwFlsIndex]) Peb->FlsCallback[dwFlsIndex](ppFlsSlots[dwFlsIndex + 2]);` ?
(I have supposed here that `ppFlsSlots[dwFlsIndex + 2]` is the lpFlsData as described in [FlsCallback](https://msdn.microsoft.com/en-us/library/windows/desktop/ms682665(v=vs.85).aspx) MSDN page.)

Reply:
Yeah but is it protected by SEH etc, and does an application rely on it?